pkg/auto/auto.go

```diff
@@ -0,0 +1,264 @@
+		return Command{}, errNoUpstreamRemote
+	}
+	commands := []string{
+		"make",
+		"forward-port",
+		"CHART=" + asset,
+		"VERSION=" + version,
+		"BRANCH=" + f.VR.DevBranch,
+		"UPSTREAM=" + upstreamRemote,
+	}
+
+	return Command{Chart: asset, Command: commands, Version: version}, nil
+}
+
+// checkIfChartChanged will check if the chart has changed from the last chart.
+// It will return true if the chart has changed, otherwise it will return false.
+// If the chart has changed, it will check for special cases like (fleet and neuvector), and CRD dependencies.
+func checkIfChartChanged(lastChart, currentChart string) bool {
+	// Check if the current chart is different from the last chart
+	if lastChart == currentChart {
+		return false
+	}
+
+	// Check for special edge cases
+	return checkEdgeCasesIfChartChanged(lastChart, currentChart)
+}
+
+// checkEdgeCasesIfChartChanged will check for special cases like:
+//
+//	-CRD dependencies
+//	-fleet
+//	-neuvector
+//	-rancher-alerting-driver + rancher-aks-operator
+//	-rancher-gke-operator + rancher-gatekeeper
+//
+// It will return true if the chart changed, false otherwise
+func checkEdgeCasesIfChartChanged(lastChart, currentChart string) bool {
+
+	lastParts := strings.Split(lastChart, "-")
+	currentParts := strings.Split(currentChart, "-")
+
+	minLength := 0 // compare which chart is shorter in name (last or current)
+	if len(currentParts) < len(lastParts) {
+		minLength = len(currentParts)
+	} else {
+		minLength = len(lastParts)
+	}
+
+	equalCounter := 0
+	for i := 0; i < minLength; i++ {
+		if lastParts[i] == currentParts[i] {
+			equalCounter++
+		}
+	}
+
+	if equalCounter == 0 {
+		return true
+	} else if equalCounter == 1 && strings.HasPrefix("rancher-", lastParts[0]) && minLength < 3 {
+		return true
+	} else if equalCounter == 1 && minLength >= 3 {
+		return true
+	}
+
+	// treat operators edge cases
+	// rancher-aks-operator && rancher-gke-operator || rancher-aks-operator-crd && rancher-gke-operator-crd
+	if (equalCounter == 2 && minLength == 3) || (equalCounter == 3 && minLength == 4) {
+		if lastParts[0] == "rancher" && currentParts[0] == "rancher" {
+			if lastParts[2] == "operator" && currentParts[2] == "operator" {
+				if lastParts[1] != currentParts[1] {
+					return true
+				}
+			}
+		}
+	}
+
+	return false
+}
+
+// createNewBranchToForwardPort will create a new branch to forward-port the assets
+func (f *ForwardPort) createNewBranchToForwardPort(branch string) error {
+	// check if git is clean and branch is up-to-date
+	err := f.git.IsClean()
+	if err != nil {
+		return err
+	}
+	// create new branch and checkout
+	return f.git.CreateAndCheckoutBranch(branch)
+}
+
+// prepareReleaseYaml will prepare the release.yaml file by erasing its content,
+// this is a good practice before releasing or forward-porting any charts with it.
+func prepareReleaseYaml() error {
+	// Check if the file exists
+	_, err := os.Stat("release.yaml")
+	if os.IsNotExist(err) {
```

Comment:
The nil check on err should always be done first and if it's not nil, then inside that scope further error checking/handling should occur.

Reply:
Fixing